File: solutions/fable-1/fable/python/brackets.py
import sys

PAIRS = {')': '(', ']': '[', '}': '{'}
OPENERS = set(PAIRS.values())


def balanced(line):
    stack = []
    for ch in line:
        if ch in OPENERS:
            stack.append(ch)
        elif ch in PAIRS:
            if not stack or stack.pop() != PAIRS[ch]:
                return False
        else:
            return False
    return not stack


def main():
    data = sys.stdin.read()
    lines = data.split('\n')
    if lines and lines[-1] == '':
        lines.pop()
    out = []
    for line in lines:
        out.append('yes' if balanced(line.rstrip('\r')) else 'no')
    sys.stdout.write('\n'.join(out) + ('\n' if out else ''))


if __name__ == '__main__':
    main()

File: solutions/fable-1/fable/javascript/brackets.js
const data = require('fs').readFileSync(0, 'utf8');
const lines = data.split('\n');
if (lines.length && lines[lines.length - 1] === '') lines.pop();

const pairs = { ')': '(', ']': '[', '}': '{' };
const out = [];

for (let raw of lines) {
  const line = raw.replace(/\r$/, '');
  const stack = [];
  let ok = true;
  for (const ch of line) {
    if (ch === '(' || ch === '[' || ch === '{') {
      stack.push(ch);
    } else if (ch in pairs) {
      if (stack.pop() !== pairs[ch]) { ok = false; break; }
    } else {
      ok = false;
      break;
    }
  }
  out.push(ok && stack.length === 0 ? 'yes' : 'no');
}

process.stdout.write(out.join('\n') + (out.length ? '\n' : ''));
